src/backend/access/heap/visibilitymap.c

```diff
@@ -662,6 +663,24 @@ vm_extend(Relation rel, BlockNumber vm_nblocks)
 
 		PageInit((Page) page, BLCKSZ, 0);
 		PageSetChecksumInplace(page, vm_nblocks_now);
+
+		/*
+		 * ZENITH: Log VM fork extension
+		 * This will prevent reading uninitialized pages in the normal flow
+		 * To minimize impact on WAL size, use XLOG_FPI encoding only page header
+		 * This is okay only when VM page contains no useful data
+		 */
+		if (RelationNeedsWAL(rel))
```

Comment:
I think this should be inside the loop above.

Reply:
are you suggesting to extend VM only when relation is logged?

Reply:
Oh, never mind. I was fooled by the way github indented the code above. It seemed to me that this new code was not inside the `while (vm_nblocks_now < vm_nblocks)` loop. But it is.